File: packages/repograph/src/scorecard/engine.py
"""Scoring engine for AI-Readiness Scorecard."""

import time
import uuid
from pathlib import Path
from typing import List, Optional
import structlog

from .models import ScorecardReport, DetectorResult, Finding
from .detectors import ALL_DETECTORS
from .formatters import JSONFormatter, MarkdownFormatter

logger = structlog.get_logger(__name__)


class ScanResult:
    """Result of a scorecard scan."""

    def __init__(self, report: ScorecardReport):
        self.report = report
        self.scan_id = report.scan_id
        self.score = report.score

    def to_json(self) -> str:
        """Export report as JSON."""
        formatter = JSONFormatter()
        return formatter.format(self.report)

    def to_markdown(self) -> str:
        """Export report as Markdown."""
        formatter = MarkdownFormatter()
        return formatter.format(self.report)


class ScorecardEngine:
    """Engine for running AI-readiness scorecard scans."""

    def __init__(self, repo_path: Path, repository_id: Optional[str] = None, tenant_id: Optional[str] = None):
        """
        Initialize scorecard engine.

        Args:
            repo_path: Path to repository to scan
            repository_id: Optional repository ID from database
            tenant_id: Optional tenant ID for multi-tenancy
        """
        self.repo_path = Path(repo_path)
        self.repository_id = repository_id
        self.tenant_id = tenant_id
        self.logger = logger.bind(repo_path=str(repo_path))

        if not self.repo_path.exists():
            raise ValueError(f"Repository path does not exist: {repo_path}")

    def scan(self, detectors: Optional[List[str]] = None) -> ScanResult:
        """
        Run scorecard scan.

        Args:
            detectors: Optional list of detector names to run. If None, runs all.

        Returns:
            ScanResult with findings and score
        """
        scan_id = str(uuid.uuid4())
        start_time = time.time()

        self.logger.info("Starting scorecard scan", scan_id=scan_id)

        # Initialize report
        report = ScorecardReport(
            scan_id=scan_id,
            repository_path=str(self.repo_path),
            repository_id=self.repository_id,
            tenant_id=self.tenant_id,
            score=100,  # Will be calculated
        )

        # Determine which detectors to run
        detectors_to_run = self._get_detectors(detectors)

        # Run each detector
        for detector_class in detectors_to_run:
            detector_result = self._run_detector(detector_class)
            report.detector_results.append(detector_result)

            # Add findings to report
            for finding in detector_result.findings:
                report.add_finding(finding)

        # Calculate final score
        report.calculate_score()

        # Calculate scan metrics
        end_time = time.time()
        report.total_scan_time_ms = (end_time - start_time) * 1000
        report.files_scanned = self._count_files()

        self.logger.info(
            "Scorecard scan completed",
            scan_id=scan_id,
            score=report.score,
            total_findings=report.total_findings,
            blockers=report.blocker_count,
            warnings=report.warning_count,
            scan_time_ms=report.total_scan_time_ms,
        )

        return ScanResult(report)

    def _get_detectors(self, detector_names: Optional[List[str]]) -> List[type]:
        """
        Get detector classes to run.

        Args:
            detector_names: Optional list of detector names

        Returns:
            List of detector classes
        """
        if detector_names is None:
            return ALL_DETECTORS

        # Filter detectors by name
        detectors = []
        for detector_class in ALL_DETECTORS:
            detector = detector_class(self.repo_path)
            if detector.name in detector_names:
                detectors.append(detector_class)

        return detectors

    def _run_detector(self, detector_class: type) -> DetectorResult:
        """
        Run a single detector.

        Args:
            detector_class: Detector class to instantiate and run

        Returns:
            DetectorResult with findings
        """
        detector = detector_class(self.repo_path)

        self.logger.debug("Running detector", detector=detector.name)

        start_time = time.time()
        error = None
        findings = []

        try:
            findings = detector.detect()
        except Exception as e:
            error = str(e)
            self.logger.error(
                "Detector failed",
                detector=detector.name,
                error=error,
                exc_info=True,
            )

        end_time = time.time()
        execution_time_ms = (end_time - start_time) * 1000

        self.logger.debug(
            "Detector completed",
            detector=detector.name,
            findings_count=len(findings),
            execution_time_ms=execution_time_ms,
        )

        return DetectorResult(
            detector_name=detector.name,
            findings=findings,
            execution_time_ms=execution_time_ms,
            error=error,
        )

    def _count_files(self) -> int:
        """Count total files scanned."""
        count = 0
        extensions = {'.py', '.ts', '.tsx', '.jsx', '.js', '.md', '.json', '.yaml', '.yml', '.html', '.vue'}

        for file_path in self.repo_path.rglob('*'):
            if file_path.is_file() and file_path.suffix in extensions:
                # Use same skip logic as detectors
                skip_dirs = {
                    'node_modules', '__pycache__', '.git', 'venv', '.venv',
                    'dist', 'build', '.pytest_cache', '.mypy_cache',
                    'site-packages', '.next', 'coverage', '.tox',
                }
                parts = set(file_path.parts)
                if not (parts & skip_dirs):
                    count += 1

        return count

File: packages/aethyme/rust/crates/aethyme-quality/src/engine.rs
//! Scoring engine (port of `src/scorecard/engine.py`).

use std::path::{Path, PathBuf};
use std::time::Instant;

use crate::detectors::{Detector, all_detectors};
use crate::model::{
    DetectorResult, InspectionDetectorResult, QualityInspection, ScorecardReport, Severity,
};
use crate::snapshot::TrackedSnapshot;
use crate::util::{now_timestamps, uuid4};
use crate::walk::{count_files_skip, py_suffix, rglob_all};

/// Extensions `_count_files` counts.
const COUNT_EXTENSIONS: [&str; 11] = [
    ".py", ".ts", ".tsx", ".jsx", ".js", ".md", ".json", ".yaml", ".yml", ".html", ".vue",
];

#[derive(Debug)]
pub struct ScorecardEngine {
    pub repo_path: PathBuf,
    pub repository_id: Option<String>,
    pub tenant_id: Option<String>,
}

impl ScorecardEngine {
    /// Errors when the repository path does not exist (Python raised
    /// `ValueError` with this message).
    pub fn new(
        repo_path: &Path,
        repository_id: Option<String>,
        tenant_id: Option<String>,
    ) -> Result<Self, String> {
        if !repo_path.exists() {
            return Err(format!(
                "Repository path does not exist: {}",
                repo_path.display()
            ));
        }
        Ok(ScorecardEngine {
            repo_path: repo_path.to_path_buf(),
            repository_id,
            tenant_id,
        })
    }

    /// Run the scan. `detectors = None` runs the full registry;
    /// otherwise the registry is filtered by name — registry order is
    /// preserved and unknown names are silently ignored, exactly like
    /// the Python `_get_detectors`.
    pub fn scan(&self, detectors: Option<&[String]>) -> ScorecardReport {
        let scan_id = uuid4();
        let start = Instant::now();
        let (timestamp_iso, timestamp_display) = now_timestamps();

        let mut report = ScorecardReport::new(
            scan_id,
            self.repo_path.display().to_string(),
            self.repository_id.clone(),
            self.tenant_id.clone(),
            timestamp_iso,
            timestamp_display,
        );

        let to_run: Vec<Box<dyn Detector>> = match detectors {
            None => all_detectors(),
            Some(names) => all_detectors()
                .into_iter()
                .filter(|d| names.iter().any(|n| n == d.name()))
                .collect(),
        };

        for detector in to_run {
            let det_start = Instant::now();
            // Python wrapped `detector.detect()` in a try/except that
            // stored the exception message in `error`. The Rust
            // detectors are total (IO failures degrade to "unreadable
            // file", like `read_file_safe`), so `error` is always None.
            let findings = detector.detect(&self.repo_path);
            let execution_time_ms = det_start.elapsed().as_secs_f64() * 1000.0;
            let result = DetectorResult {
                detector_name: detector.name().to_string(),
                findings,
                execution_time_ms,
                error: None,
            };
            for finding in &result.findings {
                report.add_finding(finding.clone());
            }
            report.detector_results.push(result);
        }

        report.calculate_score();
        report.total_scan_time_ms = start.elapsed().as_secs_f64() * 1000.0;
        report.files_scanned = self.count_files();
        report
    }

    /// Inspect only tracked, relevant repository files. Unlike the legacy
    /// scorecard, this surface records detector applicability and does not
    /// interpret quality findings as operational readiness blockers.
    pub fn inspect_tracked(
        &self,
        detectors: Option<&[String]>,
    ) -> Result<QualityInspection, String> {
        let start = Instant::now();
        let snapshot = TrackedSnapshot::materialize(&self.repo_path)?;
        let to_run: Vec<Box<dyn Detector>> = match detectors {
            None => all_detectors(),
            Some(names) => all_detectors()
                .into_iter()
                .filter(|detector| names.iter().any(|name| name == detector.name()))
                .collect(),
        };

        let mut findings = Vec::new();
        let mut detector_results = Vec::new();
        for detector in to_run {
            let applicability = detector.applicability(snapshot.root());
            let det_start = Instant::now();
            let detector_findings = if applicability.is_applicable() {
                detector.inspect(snapshot.root())
            } else {
                Vec::new()
            };
            findings.extend(detector_findings.iter().cloned());
            detector_results.push(InspectionDetectorResult {
                detector_name: detector.name().to_string(),
                description: detector.description().to_string(),
                applicability,
                findings: detector_findings,
                execution_time_ms: det_start.elapsed().as_secs_f64() * 1000.0,
                error: None,
            });
        }

        let high_count = findings
            .iter()
            .filter(|finding| finding.severity == Severity::Blocker)
            .count();
        let medium_count = findings
            .iter()
            .filter(|finding| finding.severity == Severity::Warning)
            .count();
        let low_count = findings
            .iter()
            .filter(|finding| finding.severity == Severity::Info)
            .count();

        Ok(QualityInspection {
            repository_path: snapshot.source_root().display().to_string(),
            tracked_file_count: snapshot.tracked_file_count,
            relevant_file_count: snapshot.relevant_file_count,
            excluded_vendored_count: snapshot.excluded_vendored_count,
            excluded_generated_count: snapshot.excluded_generated_count,
            excluded_non_regular_count: snapshot.excluded_non_regular_count,
            total_findings: findings.len(),
            high_count,
            medium_count,
            low_count,
            findings,
            detector_results,
            total_scan_time_ms: start.elapsed().as_secs_f64() * 1000.0,
        })
    }

    /// Port of `_count_files`: every file under the repo whose Python
    /// `suffix` is in the counted set and whose path shares no
    /// component with the skip-dir list (note: NO hidden-component
    /// rule here — the engine's check differed from the detectors').
    fn count_files(&self) -> i64 {
        let mut count = 0i64;
        for entry in rglob_all(&self.repo_path) {
            if !entry.is_file {
                continue;
            }
            let suffix = py_suffix(&entry.path);
            if !COUNT_EXTENSIONS.contains(&suffix.as_str()) {
                continue;
            }
            if count_files_skip(&entry.path) {
                continue;
            }
            count += 1;
        }
        count
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    #[test]
    fn rejects_missing_path() {
        let err = ScorecardEngine::new(Path::new("/nonexistent/path"), None, None).unwrap_err();
        assert!(err.contains("does not exist"));
    }

    #[test]
    fn counts_files_like_python() {
        let tmp = std::env::temp_dir().join(format!("aq-count-{}", std::process::id()));
        let _ = fs::remove_dir_all(&tmp);
        fs::create_dir_all(tmp.join("src")).unwrap();
        fs::create_dir_all(tmp.join("dist")).unwrap();
        fs::create_dir_all(tmp.join(".hidden")).unwrap();
        fs::write(tmp.join("a.py"), "").unwrap();
        fs::write(tmp.join("b.rs"), "").unwrap(); // not counted (.rs)
        fs::write(tmp.join("src/c.md"), "").unwrap();
        fs::write(tmp.join("dist/d.js"), "").unwrap(); // skip dir
        fs::write(tmp.join(".hidden/e.py"), "").unwrap(); // counted! no hidden rule
        let engine = ScorecardEngine::new(&tmp, None, None).unwrap();
        assert_eq!(engine.count_files(), 3);
        let _ = fs::remove_dir_all(&tmp);
    }

    #[test]
    fn empty_registry_scan_scores_100() {
        let tmp = std::env::temp_dir().join(format!("aq-scan-{}", std::process::id()));
        let _ = fs::remove_dir_all(&tmp);
        fs::create_dir_all(&tmp).unwrap();
        let engine = ScorecardEngine::new(&tmp, Some("rid".into()), Some("tid".into())).unwrap();
        let report = engine.scan(Some(&["no-such-detector".to_string()]));
        assert_eq!(report.score, 100);
        assert_eq!(report.total_findings, 0);
        assert!(report.detector_results.is_empty());
        assert_eq!(report.repository_id.as_deref(), Some("rid"));
        assert_eq!(report.tenant_id.as_deref(), Some("tid"));
        let _ = fs::remove_dir_all(&tmp);
    }

    #[test]
    fn tracked_inspection_ignores_untracked_irrelevant_ui_files() {
        use std::process::Command;

        let tmp =
            std::env::temp_dir().join(format!("aq-tracked-inspection-{}", std::process::id()));
        let _ = fs::remove_dir_all(&tmp);
        fs::create_dir_all(tmp.join("src")).unwrap();
        assert!(
            Command::new("git")
                .arg("init")
                .arg("-q")
                .arg(&tmp)
                .status()
                .unwrap()
                .success()
        );
        fs::write(tmp.join("src/lib.rs"), "pub fn answer() -> u8 { 42 }\n").unwrap();
        assert!(
            Command::new("git")
                .arg("-C")
                .arg(&tmp)
                .args(["add", "src/lib.rs"])
                .status()
                .unwrap()
                .success()
        );
        fs::write(
            tmp.join("scratch.tsx"),
            "export const X = () => <button>go</button>;\n",
        )
        .unwrap();

        let engine = ScorecardEngine::new(&tmp, None, None).unwrap();
        let report = engine
            .inspect_tracked(Some(&["data-ui-coverage".to_string()]))
            .unwrap();
        assert_eq!(report.total_findings, 0);
        assert_eq!(report.tracked_file_count, 1);
        assert_eq!(report.relevant_file_count, 1);
        assert!(matches!(
            report.detector_results[0].applicability,
            crate::detectors::DetectorApplicability::NotApplicable { .. }
        ));

        let _ = fs::remove_dir_all(&tmp);
    }

    #[test]
    fn legacy_scan_keeps_irrelevant_monorepo_findings_for_compatibility() {
        let tmp = std::env::temp_dir().join(format!(
            "aq-legacy-irrelevant-monorepo-{}",
            std::process::id()
        ));
        let _ = fs::remove_dir_all(&tmp);
        fs::create_dir_all(tmp.join("src")).unwrap();
        fs::write(
            tmp.join("Cargo.toml"),
            "[package]\nname = \"rust-only\"\nversion = \"0.1.0\"\n",
        )
        .unwrap();
        fs::write(tmp.join("src/lib.rs"), "pub fn answer() -> u8 { 42 }\n").unwrap();

        // The legacy scorecard walks arbitrary repository files, including an
        // incidental untracked UI scratch file in an otherwise Rust project.
        fs::write(
            tmp.join("scratch.tsx"),
            "export const X = () => <button>go</button>;\n",
        )
        .unwrap();

        let engine = ScorecardEngine::new(&tmp, None, None).unwrap();
        let report = engine.scan(Some(&["data-ui-coverage".to_string()]));
        assert_eq!(report.warning_count, 1);
        assert_eq!(report.warnings[0].file_path, "scratch.tsx");

        let _ = fs::remove_dir_all(&tmp);
    }
}
